rust/theoros/src/types/calldata.rs

```diff
@@ -35,72 +32,69 @@ pub struct Calldata {
-
-        let update = match event.update {
-            DispatchUpdate::SpotMedian { update, feed_id: _ } => update,
+    pub async fn build_from(state: &AppState, chain_name: EvmChainName, feed_id: String) -> anyhow::Result<Calldata> {
+        let feed_id = hex_str_to_u256(&feed_id)?;
+        let update_info = state.storage.latest_update_per_feed().get(&feed_id).await?.context("No update found")?;
+
+        let validator_index_map =
+            state.hyperlane_validators_mapping.get_validators(&chain_name).context("No validators found")?;
+
+        let validators: Vec<Felt> = validator_index_map.keys().copied().collect();
+        let checkpoints = state.storage.signed_checkpoints().get(&validators, update_info.nonce).await;
+        anyhow::ensure!(!checkpoints.is_empty(), "No signatures found");
+
+        // Ensure all nonce have the same checkpoint
+        let nonce_checkpoint = &checkpoints[0].1.value;
+        anyhow::ensure!(
+            checkpoints.iter().all(|(_, checkpoint)| &(checkpoint.value) == nonce_checkpoint),
+            "Inconsistent checkpoint values found"
+        );
+
+        let signatures: Vec<ValidatorSignature> = checkpoints
+            .iter()
+            .filter_map(|(validator, signed_checkpoint)| {
+                validator_index_map
+                    .get(validator)
+                    .map(|&idx| ValidatorSignature { validator_index: idx, signature: signed_checkpoint.signature })
+            })
+            .collect();
+
+        let update = match update_info.update {
+            DispatchUpdate::SpotMedian { update, .. } => update,
         };
 
         let payload = Payload {
-            checkpoint: checkpoint_infos.value.clone(),
+            checkpoint: nonce_checkpoint.clone(),
             num_updates: 1,
-            update_data_len: 1,
+            // TODO: proof should be deleted
             proof_len: 0,
             proof: vec![],
+            update_data_len: update.to_bytes().len() as u16,
             update_data: update.to_bytes(),
-            // TODO: Store directly a U256.
-            feed_id: U256::from_str(&feed_id).unwrap(),
+            feed_id,
+            // TODO: publish_time is a duplicated of update timestamp - remove?
             publish_time: update.metadata.timestamp,
         };
 
         let hyperlane_message = HyperlaneMessage {
             hyperlane_version: HYPERLANE_VERSION,
-            // TODO: signers_len & signatures should work for multiple validators
-            signers_len: 1_u8,
-            signatures: vec![ValidatorSignature { validator_index: 0, signature: checkpoint_infos.signature }],
-            nonce: event.nonce,
+            emitter_chain_id: update_info.emitter_chain_id,
+            emitter_address: update_info.emitter_address,
+            nonce: update_info.nonce,
+            signers_len: signatures.len() as u8,
+            signatures,
+            // TODO: timestamp is a duplicated of update timestamp - remove?
```

Comment:
it's a different timestamp technically it should be the timestamp at which the message is signed I guess ?

Reply:
although not sure if really useful

Reply:
Yep, let's talk about this later with @JordyRo1 to be definitely sure about what we want to keep there